api/envoy/config/filter/network/http_connection_manager/v2/http_connection_manager.proto

```diff
@@ -419,6 +425,115 @@ message Rds {
   string route_config_name = 2 [(validate.rules).string.min_bytes = 1];
 }
 
+// This message is used to work around the limitations with 'oneof' and repeated fields.
+message ScopedRouteConfigurationsList {
+  repeated envoy.api.v2.ScopedRouteConfiguration scoped_route_configurations = 1
+      [(validate.rules).repeated .min_items = 1];
+}
+
+message ScopedRoutes {
+  // The name assigned to the scoped routing configuration.
+  string name = 1 [(validate.rules).string.min_bytes = 1];
+
+  // Specifies the mechanism for constructing "scope keys" based on HTTP request attributes. These
+  // keys are matched against a set of :ref:`Key<envoy_api_msg_ScopedRouteConfiguration.Key>`
+  // objects assembled from :ref:`ScopedRouteConfiguration<envoy_api_msg_ScopedRouteConfiguration>`
+  // messages distributed via SRDS (the Scoped Route Discovery Service) or assigned statically via
+  // :ref:`scoped_route_configurations_list<envoy_api_field_config.filter.network.http_connection_manager.v2.ScopedRoutes.scoped_route_configurations_list>`.
+  //
+  // Upon receiving a request's headers, the Router will build a key using the algorithm specified
+  // by this message. This key will be used to look up the routing table (i.e., the
+  // :ref:`RouteConfiguration<envoy_api_msg_RouteConfiguration>`) to use for the request.
+  message ScopeKeyBuilder {
+    // Specifies the mechanism for constructing key fragments which are composed into scope keys.
+    message FragmentBuilder {
+      // Specifies how the value of a header should be extracted.
+      // The following example maps the structure of a header to the fields in this message.
+      //
+      // .. code::
+      //
+      //    X-Header: a=b;c=d
+      //    |         || |
+      //    |         || \----> <element_separator>
+      //    |         ||
+      //    |         |\----> <element.separator>
+      //    |         |
+      //    |         \----> <element.key>
+      //    |
+      //    \----> <name>
+      //
+      //    Each 'a=b' key-value pair constitutes an 'element' of the header field.
+      message HeaderValueExtractor {
+        // The name of the header field to extract the value from.
+        string name = 1 [(validate.rules).string.min_bytes = 1];
+
+        // The element separator (e.g., ';' separates 'a;b;c;d').
+        string element_separator = 2;
```

Comment:
The separator-centric extractor that comes included in SRDS is pretty specific. Do you want to add some discussion of how this might be generalized for other use cases?

Reply:
This could certainly become a `Core` type like `HeaderValue`, etc. Aside from being part of the ScopedRoutes message, how else do you see it as SRDS specific?

Reply:
I'm saying it's quite specific to the use case of SRDS you have in mind, In general, you might want to match on a regex group (although for performance reasons and bounded execution I'd discourage). But, you might want to use different separation/escaping schemes, you might want to parse out an IP address from an IP:port combination inside a list, you might want to support JSON maps, etc. Not sure how realistic these all are, but that's the lines I was thinking along.

Reply:
Yep, agreed. This is the reason why I already defined a `oneof extract_type` below. When these new use cases come  up, `RegexExtractor`, `JsonMapExtractor`, etc could be defined to handle them.